torchquad/integration/utils.py

```diff
@@ -1,53 +1,354 @@
+autoray, which are registered when importing this file
+"""
+import sys
+from pathlib import Path
 
-import torch
+# Change the path to import from the parent folder.
+# A relative import currently does not work when executing the tests.
+sys.path.append(str(Path(__file__).absolute().parent.parent))
+
+from autoray import numpy as anp
+from autoray import infer_backend, get_dtype_name, register_function
+from functools import partial
 from loguru import logger
 
+# from ..utils.set_precision import _get_precision
+from utils.set_precision import _get_precision
+
 
 def _linspace_with_grads(start, stop, N, requires_grad):
     """Creates an equally spaced 1D grid while keeping gradients
     in regard to inputs.
     Args:
-        start (torch.tensor): Start point (inclusive).
-        stop (torch.tensor): End point (inclusive).
-        N (torch.tensor): Number of points.
-        requires_grad (bool): Indicates if output should be recorded for backpropagation.
+        start (backend tensor): Start point (inclusive).
+        stop (backend tensor): End point (inclusive).
+        N (int): Number of points.
+        requires_grad (bool): Indicates if output should be recorded for backpropagation in Torch.
     Returns:
-        torch.tensor: Equally spaced 1D grid
+        backend tensor: Equally spaced 1D grid
     """
+    # The requires_grad case is only needed for Torch.
     if requires_grad:
         # Create 0 to 1 spaced grid
-        grid = torch.linspace(0, 1, N)
+        grid = anp.linspace(
+            anp.array(0.0, like=start), anp.array(1.0, like=start), N, dtype=start.dtype
+        )
 
         # Scale to desired range, thus keeping gradients
         grid *= stop - start
         grid += start
 
         return grid
     else:
-        return torch.linspace(start, stop, N)
+        if infer_backend(start) == "tensorflow":
+            # Tensorflow determines the dtype automatically and doesn't support
+            # the dtype argument here
+            return anp.linspace(start, stop, N)
+        return anp.linspace(start, stop, N, dtype=start.dtype)
+
+
+def _add_at_indices(target, indices, source, is_sorted=False):
+    """
+    Add source[i] to target at target[indices[i]] for each index i in-place.
+    For example, with targets=[0,0,0] indices=[2,1,1,2] and source=[a,b,c,d],
+    targets will be changed to [0,b+c,a+d].
+    This function supports only numpy and torch.
 
+    Args:
+        target (backend tensor): Tensor to which the source values are added
+        indices (int backend tensor): Indices into target for each value in source
+        source (backend tensor): Values which are added to target
+        is_sorted (bool, optional): Set this to True if indices is monotonically increasing to skip a redundant sorting step with the numpy backend. Defaults to False.
+    """
+    backend = infer_backend(target)
+    if backend == "torch":
+        target.scatter_add_(dim=0, index=indices, src=source)
+    else:
```

Comment:
The function might work with backends which are not JAX, TensorFlow, NumPy or PyTorch, but I haven't tested this.
I've changed the function so that it only allows torch and numpy now.